core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java

```diff
@@ -2956,7 +2963,12 @@ public static Object structAccess(Object structObject, int index, String fieldNa
     } else {
       Class<?> beanClass = structObject.getClass();
       try {
-        Field structField = beanClass.getDeclaredField(fieldName);
+        Field structField;
+        if (index >= 0 && index < beanClass.getDeclaredFields().length) {
+          structField = beanClass.getDeclaredFields()[index];
+        } else {
+          structField = beanClass.getDeclaredField(fieldName);
+        }
```

Comment:
Is there a reason of why access by index is preferred over access by name? `getDeclaredFields()` might be less efficient than `getDeclaredField()`; both involve a copy but the first one will do a copy of all fields.